### PR TITLE
Event Metrics and Validation

### PR DESCRIPTION
**Pull request by iainjwhiteside**
## Description of change

Extends the ReSim Metrics schema with a list of events, which can reference metrics defined in the metrics collection. This PR also introduces the required changes to the validator to support event validation.

the metrics SDK is updated to support events. 

## Guide to reproduce test results.

```
bazel test //...
```
## Checklist:
- [X] I have self-reviewed this change.
- [X] I have tested this change.
- [X] This change is covered by tests that are already landed, or in this PR.

### CODE REVIEW

File: resim/metrics/proto/metrics.proto [hunk 1]
```diff
@@ -224,6 +230,39 @@ message Metric {
...
     optional float order = 11;
+
+    // An optional boolean to be set to true in the case that this metric
+    // should not be displayed outside the context of the event.
+    optional bool event_metric = 12;
+}
+
+// An event is a tagged timestamp, reflecting, well, what it sounds like:
+// something happened. That event could be good, bad, neutral (and thus contains
+// the usual statues), but by definition is something worth looking at, at
```
Comment (mikebauer) [inline resim/metrics/proto/metrics.proto:241]: 🗽 ?
```suggestion
// the usual statutes), but by definition is something worth looking at, at
```

File: resim/metrics/proto/metrics.proto [hunk 2]
```diff
@@ -224,6 +230,39 @@ message Metric {
...
+// provides an insight. Thus, events, provide a way to index into the experience
```
Comment (mikebauer) [inline resim/metrics/proto/metrics.proto:242]: [nit] Not sure about the wording on this
> at provides an insight.
and
```suggestion
// the usual statues), but by definition is something worth looking at, at
// provides an insight. Thus, events provide a way to index into the experience
```

File: resim/metrics/proto/metrics.proto [hunk 3]
```diff
@@ -224,6 +230,39 @@ message Metric {
...
+// to find what matters.
+//
+// Events are associated with metrics, and can be used to provide insight into
+// the nature of the event.
```
Comment (mikebauer) [inline resim/metrics/proto/metrics.proto:246]:
```suggestion
// Events are associated with metrics, and these metrics can be used to provide insight into
// the nature of the event.
```

File: resim/metrics/python/metrics_utils.py [hunk 4]
```diff
@@ -97,6 +97,10 @@ def pack_uuid_to_proto(uuid_obj: uuid.UUID) -> uuid_pb2.UUID:
     uuid_msg.data = str(uuid_obj)
     return uuid_msg
 
+def pack_uuid_to_metric_id (uuid_obj: uuid.UUID) -> metrics_pb2.MetricID:
```
Comment (mikebauer) [inline resim/metrics/python/metrics_utils.py:100]: [nit]
```suggestion
def pack_uuid_to_metric_id(uuid_obj: uuid.UUID) -> metrics_pb2.MetricID:
```

File: resim/metrics/python/metrics_writer.py [hunk 5]
```diff
@@ -57,6 +60,16 @@ def add_metric(self, metric: MetricT) -> MetricT:
...
 
+    def base_add_event(self, event: Event) -> Event:
+        assert event.name not in self.names
+        self.names.add(event.name)
+        self.events[event.id] = event
+        return event
+
+    def add_event(self, name: str) -> Event:
+        event = Event(name=name)
+        return self.base_add_event(event)
```
Comment (mikebauer) [inline resim/metrics/python/metrics_writer.py:71]: [nit] Would you mind adding some docstrings here? I think we should have docstrings on a lot more of these methods, but in this case, it was initially confusing to me when you might use the base_ variant vs. the standard one.

**Review by mikebauer — APPROVED** [review]
Nice! These changes look good to me!

File: resim/metrics/proto/metrics.proto (hunk 1 above)
Comment (iainjwhiteside) [reply resim/metrics/proto/metrics.proto:241]: statuses 🤦

File: resim/metrics/python/metrics_utils.py (hunk 4 above)
Comment (iainjwhiteside) [reply resim/metrics/python/metrics_utils.py:100]: bad pylint, bad!

File: resim/metrics/python/metrics_writer.py (hunk 5 above)
Comment (iainjwhiteside) [reply resim/metrics/python/metrics_writer.py:71]: yeah, good point

**Pull request merged**